steam_buddy/streaming.py

```diff
@@ -0,0 +1,84 @@
+import os
+import shlex
+import time
+import subprocess as sp
+
+
+class StreamServer:
+
+    def __init__(self, sls_conf_file):
+        self.sls_conf_file = sls_conf_file
+        self._ffmpeg_cmd = "ffmpeg"
+        self._sls_cmd = "sls"
+        self._sls = None
+        self._ffmpeg = None
+
+    def __generate_sls_conf(self):
+        pass
+
+    def __start_sls(self):
+        if self._sls is None:
+            self._sls = sp.Popen([self._sls_cmd, '-c', self.sls_conf_file])
+        else:
+            raise(Exception("Error starting SLS: Already started"))
+
+    def __stop_sls(self):
+        self._sls.terminate()
+        try:
+            self._sls.wait(5)
+        except sp.TimeoutExpired:
+            self._sls.kill()
+        self._sls = None
+
+    def __start_ffmpeg(self, local=False):
+        command_line = ""
+        VCODEC = "libx264"
+        VCODEC_OPTIONS = "-preset ultrafast -tune zerolatency"
+        VSIZE = "1920x1080"
```

Comment:
Could this be based on the current resolution instead or would that be too heavy for higher resolutions?

Reply:
There are two resolutions: the source resolution and the output resolution. If you do not add the `-video_size` option then the output is based on the source. It will be better from an encoder hover head standpoint to have them both the same but the problem becomes in the server side and network bandwidth and stability. Since we are not making multiple streams with different resolutions this parameter will balance that tradeoff.
My setup's current resolution is in 4k@60Hz and most games can keep it up in ultra settings. But the streaming part adds some overhead that will drop streaming fps to about 50. Network bandwidth also suffers from large frames and high rate. It's a balancing act we have to make configurables and available to the user IMO. The recording and transcoding is intended to be almost uncompressed for faster capture and low overhead on the host PC. Then whatever captures the stream can do whatever to it.

Maybe add some logic to add or not the `-video_size` or just make it equal to the source.